dwa_local_planner/src/dwa_planner.cpp

```diff
@@ -252,8 +259,87 @@ namespace dwa_local_planner {
     return false;
   }
 
+  uint32_t DWAPlanner::cropPlanToReachableGoal(
+    const geometry_msgs::PoseStamped& global_pose,
+    const std::vector<geometry_msgs::Point>& footprint_spec,
+    std::vector<geometry_msgs::PoseStamped>& plan) {
+    // search the path for a reachable goal
+    const double dist_to_goal = mbf_utility::distance(global_pose, plan.back());
+
+    // find furthest point along the path of which the corresponding cell is in free space
+    const costmap_2d::Costmap2D* const costmap = planner_util_->getCostmap();
+    int last_point_on_the_map = -1;
+    int last_point_in_free_space = -1;
+    for (unsigned int i = 0; i < plan.size(); ++i) {
+      double g_x = plan[i].pose.position.x;
+      double g_y = plan[i].pose.position.y;
+      unsigned int map_x, map_y;
+      if (!costmap->worldToMap(g_x, g_y, map_x, map_y)) {
+        // give up once a point within the map had been found
+        if (last_point_on_the_map >= 0) {
+          break;
+        }
+        continue;
+      }
 
-  void DWAPlanner::updatePlanAndLocalCosts(
+      last_point_on_the_map = i;
+      if (costmap->getCost(map_x, map_y) < costmap_2d::INSCRIBED_INFLATED_OBSTACLE) {
+        last_point_in_free_space = i;
+      }
+    }
+
+    // handle general errors
+    if (last_point_on_the_map == -1) {
+      ROS_ERROR("None of the points of the global plan were in the local costmap, global plan points too far from robot");
+      return mbf_msgs::ExePathResult::OUT_OF_MAP;
+    }
+
+    if (last_point_in_free_space == -1) {
+      ROS_ERROR("None of the points of the global plan were in free space");
+      return mbf_msgs::ExePathResult::BLOCKED_PATH;
+    }
+
+    const auto dist_to_last_free_point = mbf_utility::distance(global_pose, plan[last_point_in_free_space]);
+    const auto limits = planner_util_->getCurrentLimits();
+    if (dist_to_last_free_point > limits.blocked_goal_approach_distance) {
+      // -> just crop the path to this point
+      plan.resize(last_point_in_free_space+1);
+      return mbf_msgs::ExePathResult::SUCCESS;
+    }
```

Comment:
Q: What's happening here? From the parameter description for `blocked_goal_approach_distance` it seems like we would approach the goal until this distance (i.e. similar to your previous feature of stopping withing X meter of the goal). But in this `if statement` it seems more like you're treating this parameter as the distance from the start pose?

Reply:
So what's happening here is that we search for a point on the path in free space; so if we find one more than `blocked_goal_approach_distance`, we're done.

So as a result, if the goal is blocked, we continue moving if the last point before the goal blockage is more than `blocked_goal_approach_distance` away.

But yes, it is technically possible that this last free point on the path is further away than the goal; `blocked_goal` in `blocked_goal_approach_distance` basically refers to the point where the blockage starts, not where the goal is.
(was also implemented like this in the previous feature you were referring to)

So... any suggestions for improving the parameter name? Maybe `goal_blockage_approach_distance`? `goal_obstacle_approach_distance`?

Reply:
How about something like `goal_obstacle_min_dist`?

Reply:
so far my preference would be `goal_obstacle_approach_distance`, it is more clear to me what that means

Reply:
renamed in https://github.com/rapyuta-robotics/ros-navigation/pull/74/commits/0b895adcd69e3aab19f38f5c69373ca5354f72e5;
I also took the liberty to update its description based on the comments here (https://github.com/rapyuta-robotics/ros-navigation/pull/74/commits/d7bb99dadf4f189c1a1c077f032f38d219e2076c) such that it is hopefully a bit more clear what the parameter does